### PR TITLE
[WIP] Opt-in for smooth pointer scrolling

### LINKED ISSUE #32120: Scroll wheel scrolling is not smooth
## Steps to Reproduce

1. Use a desktop embedder, or another platform that allows you to scroll with a mouse's scroll wheel.
2. Scroll any scrollable content (e.g. list demo in Flutter Gallery)
3. Observe how the scroll offset abruptly jumps between scroll notches instead of smoothly transitioning (like Chrome, Firefox, UWP apps, etc.)

Gif example:

### PR DESCRIPTION
**Pull request by Piinks**
WIP

Fixes https://github.com/flutter/flutter/issues/32120

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/widgets/nested_scroll_view.dart [hunk 1]
```diff
@@ -899,12 +899,22 @@ class _NestedScrollCoordinator implements ScrollActivityDelegate, ScrollHoldCont
...
 
-  void pointerScroll(double delta) {
+  void pointerScroll(double delta, bool animatePointerScroll) {
     // If an update is made to pointer scrolling here, consider if the same
     // (or similar) change should be made in
     // ScrollPositionWithSingleContext.pointerScroll.
     assert(delta != 0.0);
 
+    if (animatePointerScroll) {
+      double currentPosition = _outerPosition?.pixels ?? 0.0;
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/nested_scroll_view.dart:909]: Another point for https://github.com/flutter/flutter/issues/102003, this kind of info should be more accessible.

File: packages/flutter/lib/src/widgets/nested_scroll_view.dart [hunk 2]
```diff
@@ -899,12 +899,22 @@ class _NestedScrollCoordinator implements ScrollActivityDelegate, ScrollHoldCont
...
+      for (final _NestedScrollPosition innerPosition in _innerPositions) {
+        currentPosition += innerPosition.pixels;
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/nested_scroll_view.dart:911]: TODO(me): fix this.

**Review by yjbanov — COMMENTED** [review]
From some manual Q/A:

* If scroll wheel event are high enough frequency, the scrolling velocity is reset to zero on every click.
* In the Gallery home page, the horizontally scrollable carousel of demos overscrolls the target item then scrolls back to it. This happens when using the arrow button to scroll. I don't think we support horizontal scrolling using the mouse wheel (?), so I didn't test it.

**Comment by Piinks** [comment]
> If scroll wheel event are high enough frequency, the scrolling velocity is reset to zero on every click.

Is this a case where the scroll direction is changing? I recall you mentioned testing this with the Gallery app, did you opt-in for each individual demo? I think you said you saw it in action on the home page, if you set MaterialApp.scrollBehavior to opt in at that level, the demoes with their own MaterialApps override it. They each need to set it in the Gallery app since they are designed as stand alone apps within the gallery app.

File: packages/flutter/lib/src/widgets/scroll_position_with_single_context.dart [hunk 3]
```diff
@@ -208,31 +209,81 @@ class ScrollPositionWithSingleContext extends ScrollPosition implements ScrollAc
...
+      // Compute the delta-based duration for the new input
+      final double newDuration = physics.getPointerAnimationDurationFor(delta);
+      final double carriedVelocity = 1.3 / (_pointerScrollActivity!.velocity.clamp(-1000, 1000) * 0.42);
+      print('Continue animating, carried velocity control point: $carriedVelocity, duration: $newDuration');
+      _pointerScrollActivity = DrivenScrollActivity(
+        this,
+        from: pixels,
+        to: newTargetPixels,
+        duration: Duration(milliseconds: newDuration.round()),
+        curve: Cubic(0.42, carriedVelocity, 0.58, 1.0),
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/scroll_position_with_single_context.dart:281]: @yjbanov this is the updated curve

**Review by Piinks — COMMENTED** [review]
@yjbanov now that https://github.com/flutter/engine/pull/36346 has landed, can you try this again? Since we can now identify trackpads, I have filtered them out. :)

**Comment by Piinks** [comment]
Closing for now since there has been no feedback. Will look at returning to this in a few weeks.

**Comment by DexEze** [comment]
I wonder if you can pick thos back up @Piinks?
It would be good to have a smooth scroll on desktop and web with mouse scroll wheel

**Comment by water-mizuu** [comment]
Please look into picking this back up @Piinks 😢, as of now, I've been using my own fork of [pub.dev/smooth_scroll_multiplatform](https://pub.dev/packages/smooth_scroll_multiplatform), which breaks with NotificationListeners.

**Pull request closed without merging**